discussion/rust-for-embedded-systems.tex

```diff
@@ -0,0 +1,90 @@
+It is easy enough to define programs that are not dependent on this {\std} library, and instead only dependent on {\core}, but we think the difference between {\std} and {\core} is too large.
+
+\gls{rcl} does provide us with enough functionality to write idiomatic {\rust} code, but we had to implement a few workarounds in order to get support for {\rust}'s standard constructs for dynamic memory allocation.
+Currently, {\rust} does not have functionality to provide, or automatically compile, any of its standard libraries for target architectures other than the ones it supports by default.
+Because our project targets an ARM processor architecture, we have to conditionally download and compile the libraries (like \lib{alloc} and \lib{collections}) that have functionality that we want to utilize, as described in \autoref{ssub:transitioning_to_cargo}.
+This is not necessarily a big problem in itself, as it does not complicate the process much more than the need to explicitly define the libraries as custom dependencies in the \texttt{Cargo.toml}.
+We do, however, think that the distinction between standard and non-standard programs make it considerably harder to enable us to use the great ecosystem that {\rust} is surrounded by.
+
+\subsection{Using and Distributing Libraries}
+
+We have described {\cargo} and its ability to manage package dependencies, and a package repository closely associated with {\cargo} is called \texttt{crates.io}\footnote{\url{https://crates.io}}.
+This repository stores, at the time of this writing, a couple of thousand different {\rust} crates, all made available for anyone who builds {\rust} programs with {\cargo}.
+The vast majority of these crates are, to the best of our knowledge, linked to {\std} and utilize some part of it, either directly or transitively through a package dependency.
+That being said, they do not necessarily use any functionality that we have not already made available through \gls{rel} in our project.
+This simple fact renders nearly all (if not all) packages available through \texttt{crates.io} unusable for our project, even though the same basis that many of these packages depend on might already be available through \gls{rel}.
+
+As an example to demonstrate this problem, we look to the \texttt{lazy-static}\footnote{\url{https://crates.io/crates/lazy_static}} project available on \texttt{crates.io}.
+This project allows the programmer to declare \emph{static} variables that get initialized at \emph{runtime}.
+It provides some of the same functionality that we implemented for the event-hub described in \autoref{sec:irq-closures}.
+We can not use this project directly because it depends on {\std}, but if we look closer at the project's source code, we can see that the project author has implemented an alternate (and outdated) version which is only dependent on {\core}.
```

Comment:
```
We can not use this project directly because it depends on {\std}.  With a closer look at the project, we can see that the project author as some point did prototype an alternate (and outdated) version which was only dependent on {\core}.
```